core/src/main/java/org/infinispan/globalstate/impl/GlobalStateManagerImpl.java

```diff
@@ -0,0 +1,132 @@
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * GlobalStateManagerImpl. This global component manages persistent state across restarts. The
+ * information is stored in a Properties file. On a graceful shutdown it persists the following
+ * information:
+ *
+ * version = full version (e.g. major.minor.micro.qualifier)
+ * timestamp = timestamp using ISO-8601
+ *
+ * as well as any additional information contributed by registered {@link GlobalStateProvider}s
+ *
+ * @author Tristan Tarrant
+ * @since 8.1
+ */
+public class GlobalStateManagerImpl implements GlobalStateManager {
+   private static Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private GlobalConfiguration globalConfiguration;
+   private List<GlobalStateProvider> stateProviders = new ArrayList<>();
+   private TimeService timeService;
+
+   @Inject
+   public void inject(GlobalConfiguration globalConfiguration, TimeService timeService,
+         EmbeddedCacheManager cacheManager) {
+      this.globalConfiguration = globalConfiguration;
+      this.timeService = timeService;
+   }
+
+   @Start(priority = 1) // Must start before everything else
+   public void start() {
+      File stateFile = getStateFile(GLOBAL_SCOPE);
+      Optional<ScopedPersistentState> globalState = readState(GLOBAL_SCOPE);
+      if (globalState.isPresent()) {
+         ScopedPersistentState state = globalState.get();
+         // We proceed only if we can write to the file
+         if (!stateFile.canWrite()) {
+            throw log.nonWritableStateFile(stateFile);
+         }
+         // Validate the state before proceeding
+         log.globalStateLoad(state.getProperty("version"), state.getProperty("timestamp"));
+
+         stateProviders.forEach(provider -> provider.prepareForRestore(state));
+      } else {
+         // Clean slate. Try to create an empty state file before proceeding
+         try {
+            stateFile.getParentFile().mkdirs();
+            stateFile.createNewFile();
+         } catch (IOException e) {
+            throw log.nonWritableStateFile(stateFile);
+         }
+      }
+
+   }
+
+   @Stop(priority = 1)
+   public void stop() {
+      ScopedPersistentState state = new ScopedPersistentStateImpl(GLOBAL_SCOPE);
+      state.setProperty("version", Version.getVersion());
+      state.setProperty("timestamp", timeService.instant().toString());
+      // ask any state providers to contribute to the global state
+      stateProviders.forEach(provider -> provider.prepareForShutdown(state));
+      writeState(state);
+      log.globalStateWrite(state.getProperty("version"), state.getProperty("timestamp"));
+   }
+
+   @Override
+   public void writeState(ScopedPersistentState state) {
+      File stateFile = getStateFile(state.getScope());
+      try (PrintWriter w = new PrintWriter(stateFile)) {
+         state.forEach((key, value) -> {
+            w.printf("%s=%s%n", key, value);
```

Comment:
Does this mean a state value must not contain '\n', and a key must not contain '='? Perhaps the `ScopedPersistentState` javadoc should explicitly say what the allowed characters are?